proposals/0529-th-quotes-as-patterns.rst

```diff
@@ -0,0 +1,356 @@
+  +        exprs
+  +      param = foldr
+  +        (\(name, _) rest -> [p| HCons $(pure $ VarP name) $rest |])
+  +        [p| HNil |]
+  +        exprs
+  +  [| (\ $param -> $(pure e')) <$> distributeFHListOverDynPure $arg |]
+  -
+  -conPCompat :: Name -> [Pat] -> Pat
+  -#if MIN_VERSION_template_haskell(2, 18, 0)
+  -conPCompat name = ConP name []
+  -#else
+  -conPCompat = ConP
+  -#endif
+
+Notes that ``conPCompat`` is gone entirely!
+By using quotes and splices like this, one avoids the AST and its instability problems.
+Quotes and splices are much more stable for the same reason the surface language is.
+The overall method of this proposal is to allow using them to solve more problems, so the AST becomes less necessary to use, and thus TH code in practice is less likely to break.
+
+Just a few lines above, however, there was more TH AST usage I couldn't get rid of::
+
+  let f :: forall d. Data d => d -> StateT [(Name, Exp)] Q d
+      f d = case eqT of
+        Just (Refl :: d :~: Exp)
+          | AppE (VarE m) eInner <- d
+          , m == 'unqMarker
+          -> do n <- lift $ newName "dynamicQuotedExpressionVariable"
+                modify ((n, eInner):)
+                return $ VarE n
+        _ -> gmapM f d
+  (e', exprsReversed) <- runStateT (gmapM f e) []
+
+Perhaps we can take solace in a claim that ``AppE`` and ``VarE`` are less likely to change, but that isn't satisfactory --- what about more complex patterns?
+The fundamental problem here is that in *positive* position (expressions), we have a choice of either using regular syntax or quotes,
+but in *negative* position (patterns), we only have the option regular syntax.
+
+The solution is simple: let's allow quotes too!
+This would allow:
+
+.. code-block:: diff
+
+         Just (Refl :: d :~: Exp)
+  -        | AppE (VarE m) eInner <- d
+  +        | [e| $(VarE m) $eInner |] <- d
+           , m == 'unqMarker
+           -> do ...
+
+or even going further:
+
+.. code-block:: diff
+
+         Just (Refl :: d :~: Exp)
+  -        | AppE (VarE m) eInner <- d
+  -        , m == 'unqMarker
+  +        | [e| unqMarker $eInner |] <- d
+           -> do ...
+
+In this way, we also avoid the use of the AST.
+
+Note we do have quotes *of patterns* today (``[p| ... |]``), but that is orthogonal.
+This is quotes *as patterns*, the type of syntax being quoted doesn't matter and could be anything.
+The point is the quotes are in negative position.
+
+With this change put together, the hope is that a significant portion of TH out in the wild is going to be more stable across GHC versions.
+This is all accomplished without trying to minimize TH AST changes, which is quite a hopeless task and also a perverse incentive for the rest of language development.
+
+Proposed Change Specification
+-----------------------------
+
+Quotes as patterns
+~~~~~~~~~~~~~~~~~~
+
+With the new extension ``TemplateHaskellQuotesAsPatterns``, slightly modified quotes are usable in pattern position.
```

Comment:
I see that @int-index also had a suggestion in https://github.com/ghc-proposals/ghc-proposals/pull/529#issuecomment-1340406138, although my inclination is still to have explicit prepositions rather than relying on word order to distinguish "pattern quote" from "quote pattern". I'm happy to leave the bikeshed colour to be resolved later once the substance of the proposal is agreed, though.

I still think it would be helpful to have some explicit definitions, something like (with the current terminology):

 * A quote is said to be *of* the syntactic category being quoted, for example `[e| ... |]` is a *quote of an expression* and `[p| ... |]` is a *quote of a pattern*. Template Haskell already supports quotes of expressions, patterns, types and declarations, plus quotes of arbitrary syntax via quasiquotes.
 * A quote is said to be used *as* the syntactic category in which it occurs, for example `x = [...| ... |]` includes a  *quote as an expression* whereas `f [...| ... |] = ...` includes a *quote as a pattern*. Template Haskell currently supports quotes as expressions and quasiquotes as all four categories; this proposal adds quotes as patterns.
 * Thus in `f [t| Bool |] = [d| data T |]` we have a *quote of a type as a pattern* and a *quote of a declaration as an expression*.